### DISCUSSION TITLE
mandatory parameters

### DISCUSSION THREAD

**Discussion by aowss** [Randgalt/record-builder#32] [Q&A]
Hi,

typically a builder's constructor ( or static factory ) takes all the required parameters.
When generating the builder via Includes, is there a way to specify the required parameters ?

Thanks

**Comment by Randgalt** [comment]
Do you mean `RecordInterface.Include`? Maybe you can give an example?

**Reply by aowss** [reply]
If i have a record like that

`public record Info(@NotNull String title, String summary, String description, URL termsOfService, Contact contact, License license, @NotNull String version) {}`

and if i have a class like that

```
@RecordBuilder.Include({Info.class})
public class OpenApiBuilder {
}
```

Is there a way to specify that `title` and `version` are required and therefore get a builder that enforces that. An equivalent of

```
public Builder(String title, String version) {
            this.title = title;
            this.version = version;
}
```

**Reply by Randgalt** [reply]
I see - this relates somewhat to the discussion here: https://github.com/Randgalt/record-builder/issues/30 i.e. how should validation annotations be handled. I should see what the Immutables library does with this and maybe try to emulate it. However, I really want to keep this library simple too.

**Reply by aowss** [reply]
Thanks for your prompt response and for this library.

I tried to summarize the options in [this doc](https://github.com/aowss/open-api/tree/main/model).
What i like with this library is the ability to not annotate the model ( that is very important to me ) but having the mandatory parameters part of the generated builder constructor is also important.
I think it is not exactly the same as the discussion you reference since this is part and parcel of the builder pattern.

Let me know what you think. Whatever you decide to do, good work!

**Reply by Randgalt** [reply]
Great doc with very a thoughtful description of the issue. Bundling this with https://github.com/Randgalt/record-builder/issues/30 we get...

1. Should RecordBuilder do anything in particular with validation annotations
2. Should RecordBuilder offer specialized constructors and/or helpers beyond what the language offers for records?

For 1, we could do:

- Nothing
- Copy any annotations without further processing
- Have RecordBuilder notice common annotations such as `NotNull` and do further processing such as what the Immutables library does
- Have RecordBuilder optionally pass built objects through a JSR-303 validator

For 2, we could
- Rely on the solution to 1 above - i.e. the validation annotations are sufficient
- Do nothing
- Create alternate ctors as described in @aowss's document

**Reply by Randgalt** [reply]
See proposal here: https://github.com/Randgalt/record-builder/issues/30#issuecomment-853726813

**Reply by aowss** [reply]
I think that there are 2 approaches:

- enrich the annotation to tell it which parameter should go in the builder constructor but this opens up the possibility of mistakes + this would generate only one additional constructor with the mentioned parameters

- build one builder constructor per record constructor. At the moment you build one no-arg constructor for the record's canonical constructor. You could then add one constructor per record constructor matching the signature.

so if the [record](https://github.com/aowss/open-api/blob/main/model/src/main/java/openapi/model/v310/Info.java) was

```
public record Info(@NotNull String title, String summary, String description, URL termsOfService, Contact contact, License license, @NotNull String version) {

    public Info(@NotNull String title, @NotNull String version) {
        this(title, null, null, null, null, null, version);
    }

}
```

you would generate 2 constructors : `InfoBuilder()` and `InfoBuilder(@NotNull String title, @NotNull String version)`

But the following questions arise: 
- should you even generate `InfoBuilder()`?
- what if the alternate record constructors do not mean mandatory parameters but alternative ones ?

for example this [record](https://github.com/aowss/open-api/blob/main/model/src/main/java/openapi/model/v310/License.java):

```
public record License(@NotNull String name, String identifier, URL url) {

    public License {
        if (identifier != null && url != null) throw new IllegalArgumentException("A license can't have both an identifier and a url");
    }

    public License(@NotNull String name, String identifier) {
        this(name, identifier, null);
    }

    public License(@NotNull String name, URL url) {
        this(name, null, url);
    }

}
```

based on that i would go, if you still think it is worth it, go with the first approach but allowing multiple `@RecordBuilder` or `@@RecordBuilder.Include` annotations or allowing these annotations to take a list of mandatory parameters for each builder constructor to generate.

**Comment by Randgalt** [comment]
See https://github.com/Randgalt/record-builder/pull/40 and https://github.com/Randgalt/record-builder/pull/42
